src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/CSharp/TypeAccessibilityAddUsingCSharpCodeActionResolver.cs

```diff
@@ -0,0 +1,147 @@
+            }
+
+            if (documentVersionCache is null)
+            {
+                throw new ArgumentNullException(nameof(documentVersionCache));
+            }
+
+            _foregroundDispatcher = foregroundDispatcher;
+            _documentResolver = documentResolver;
+            _documentVersionCache = documentVersionCache;
+        }
+
+        public override string Action => LanguageServerConstants.CodeActions.AddUsing;
+
+        public async override Task<CodeAction> ResolveAsync(
+            CSharpCodeActionParams csharpParams,
+            CodeAction codeAction,
+            CancellationToken cancellationToken)
+        {
+            if (csharpParams is null)
+            {
+                throw new ArgumentNullException(nameof(csharpParams));
+            }
+
+            if (codeAction is null)
+            {
+                throw new ArgumentNullException(nameof(codeAction));
+            }
+
+            var resolvedCodeAction = await ResolveCodeActionWithServerAsync(codeAction, cancellationToken);
+            if (resolvedCodeAction.Edit?.DocumentChanges is null)
+            {
+                // Unable to resolve code action with server, return original code action
+                return codeAction;
+            }
+
+            if (resolvedCodeAction.Edit.DocumentChanges.Count() != 1)
+            {
+                // We don't yet support multi-document code actions, return original code action
+                return codeAction;
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var documentChanged = resolvedCodeAction.Edit.DocumentChanges.First();
+            if (!documentChanged.IsTextDocumentEdit)
+            {
+                // Only Text Document Edit changes are supported currently, return original code action
+                return codeAction;
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var addUsingTextEdit = documentChanged.TextDocumentEdit.Edits.FirstOrDefault();
+            if (addUsingTextEdit is null)
+            {
+                // No text edit available
+                return codeAction;
+            }
+
+            var @namespace = AddUsingsCodeActionProviderFactory.ExtractNamespaceFromVSCSharpAddUsing(addUsingTextEdit.NewText);
+            if (string.IsNullOrEmpty(@namespace))
+            {
+                // Invalid text edit, missing namespace
+                return codeAction;
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+            
+            var documentSnapshot = await Task.Factory.StartNew(() =>
+            {
+                _documentResolver.TryResolveDocument(csharpParams.RazorFileUri.GetAbsoluteOrUNCPath(), out var documentSnapshot);
+                return documentSnapshot;
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler).ConfigureAwait(false);
```

Comment:
duplicate of https://github.com/dotnet/aspnetcore-tooling/pull/2684/files#diff-657af360420b13b95580366313c3358d2eb001bc2be838eed55c40cbf5af1fd5R52 ?  Consider extracting common helper for this?

Reply:
Going to leave as is for now, as this is usually the convention we use throughout. However, I agree, helpers for these repetitive expressions would be beneficial. 

Created https://github.com/dotnet/aspnetcore/issues/27499 to track.